### PR TITLE
inplace drop partition

### PR DESCRIPTION
**Pull request by luqun**
This change is to enable inplace drop partition support for myrocks.  Major code change is to iterator all partitions and find which partition need to be dropped, call SE(myrocks) API  to drop that table with constructed partition name.

This doesn't solve DDL crash safe issue.

also steal np_inplace_vs_copy testcase from George

### CODE REVIEW

**Comment by hermanlee** [comment]
cc: @georgelorchpercona @satya-bodapati

File: sql/partitioning/partition_base.cc [hunk 1]
```diff
@@ -4046,6 +4049,31 @@ bool Partition_base::commit_inplace_alter_table(
   if (ha_alter_info->alter_info->flags == Alter_info::ALTER_PARTITION)
     DBUG_RETURN(false);
 
+  std::string full_table_path = std::string("./") + altered_table->s->db.str +
+                                std::string("/") +
+                                altered_table->s->table_name.str;
```
Comment (satya-bodapati) [inline sql/partitioning/partition_base.cc]: this looks problematic. Better use tablename_to_filename()

File: sql/partitioning/partition_base.cc [hunk 2]
```diff
@@ -4046,6 +4049,31 @@ bool Partition_base::commit_inplace_alter_table(
...
+      partition_state state = part_elem->part_state;
+      switch (state) {
+        case PART_TO_BE_DROPPED:
+          char name[FN_REFLEN];
+          // TODO(luqun): add subpartition support
+          create_partition_name(name, full_table_path.c_str(),
+                                part_elem->partition_name, false);
+          (*file)->ha_external_lock(get_thd(), F_UNLCK);
+          (*file)->ha_delete_table(name, old_table_def);
+          break;
```
Comment (satya-bodapati) [inline sql/partitioning/partition_base.cc]: how do we handle if one partition couldn't be deleted/dropped? I guess ALTER guarantees no concurrent DMLS.

File: sql/partitioning/partition_base.cc [hunk 3]
```diff
@@ -4046,6 +4049,31 @@ bool Partition_base::commit_inplace_alter_table(
...
+    while ((part_elem = part_it++) != nullptr) {
+      partition_state state = part_elem->part_state;
+      switch (state) {
+        case PART_TO_BE_DROPPED:
+          char name[FN_REFLEN];
+          // TODO(luqun): add subpartition support
+          create_partition_name(name, full_table_path.c_str(),
+                                part_elem->partition_name, false);
+          (*file)->ha_external_lock(get_thd(), F_UNLCK);
+          (*file)->ha_delete_table(name, old_table_def);
```
Comment (satya-bodapati) [inline sql/partitioning/partition_base.cc]: Please check if concurrent SELECTs can be done.  Does the delete_table() implementation in myrocks gurantee to wait until pending IO is completed?

**Review by satya-bodapati — CHANGES_REQUESTED** [review]

File: sql/partitioning/partition_base.cc [hunk 4]
```diff
@@ -4046,6 +4049,31 @@ bool Partition_base::commit_inplace_alter_table(
...
+    handler **file = m_file;
+    partition_info *part_info = ha_alter_info->modified_part_info;
+    List_iterator_fast<partition_element> part_it(part_info->partitions);
+    partition_element *part_elem;
+    while ((part_elem = part_it++) != nullptr) {
+      partition_state state = part_elem->part_state;
+      switch (state) {
+        case PART_TO_BE_DROPPED:
+          char name[FN_REFLEN];
+          // TODO(luqun): add subpartition support
```
Comment (satya-bodapati) [inline sql/partitioning/partition_base.cc]: Time to add the support for sub partition as well?

File: sql/partitioning/partition_base.cc [hunk 5]
```diff
@@ -4046,6 +4049,31 @@ bool Partition_base::commit_inplace_alter_table(
...
+          create_partition_name(name, full_table_path.c_str(),
+                                part_elem->partition_name, false);
+          (*file)->ha_external_lock(get_thd(), F_UNLCK);
```
Comment (satya-bodapati) [inline sql/partitioning/partition_base.cc]: why is the unlock necessary here?

**Review by satya-bodapati — CHANGES_REQUESTED** [review]

File: sql/partitioning/partition_base.cc (hunk 2 above)
Comment (luqun) [reply sql/partitioning/partition_base.cc]: Yes, SQL layer will acquire exclusive lock and clean table cache before commit

https://github.com/facebook/mysql-5.6/blob/5dd1adebbcc81a8db2c8c30eb20c63d484ee68de/sql/sql_table.cc#L12648

File: sql/partitioning/partition_base.cc (hunk 5 above)
Comment (luqun) [reply sql/partitioning/partition_base.cc]: handler::ha_delete_table requires table/partition is unlocked..
https://github.com/facebook/mysql-5.6/blob/fb-mysql-8.0.17/sql/handler.cc#L4962

also table/partition is locked in 
https://github.com/facebook/mysql-5.6/blob/fb-mysql-8.0.17/sql/sql_table.cc#L12516

File: sql/partitioning/partition_base.cc (hunk 1 above)
Comment (luqun) [reply sql/partitioning/partition_base.cc]: Good point.. Use table->s->normalized_path.str instead.

File: sql/partitioning/partition_base.cc (hunk 4 above)
Comment (luqun) [reply sql/partitioning/partition_base.cc]: Done

File: sql/partitioning/partition_base.cc (hunk 3 above)
Comment (luqun) [reply sql/partitioning/partition_base.cc]: myrocks is "append only" database and there is no background threads to service IO request. AFAIK, it doesn't need to wait for pending IO.

Comment (luqun) [reply sql/partitioning/partition_base.cc]: > Please check if concurrent SELECTs can be done. 

there are two small time windows drop partition will acquire/hold exclusive table lock, During these two time windows, SELECT will wait for acquire table lock if it doesn't hold table lock yet.

**Comment by luqun** [comment]
@georgelorchpercona @satya-bodapati Thanks a lot for your comments, would you take a look again?

**Comment by george-lorch** [comment]
Sure @luqun I will pull it, take a look, and test it again.

**Pull request closed without merging**